File: TsinghuaNetHelper/LoginTask.cpp
#include "pch.h"

#include "LoginTask.h"

using namespace winrt;
using namespace Windows::ApplicationModel::Background;

namespace winrt::TsinghuaNetHelper::implementation
{
    fire_and_forget LoginTask::Run(IBackgroundTaskInstance const taskInstance)
    {
        auto deferral{ taskInstance.GetDeferral() };
        if (settings.AutoLogin())
        {
            try
            {
                hstring ssid;
                auto status{ SettingsHelper::InternetStatus(ssid) };
                NetState state{ settings.SuggestNetState(status, ssid) };
                hstring un{ settings.StoredUsername() };
                hstring pw{ CredentialHelper::GetCredential(un) };
                IConnect helper{ ConnectHelper::GetHelper(state, un, pw) };
                if (helper)
                {
                    co_await helper.LoginAsync();
                    FluxUser user{ co_await helper.FluxAsync() };
                    NotificationHelper::UpdateTile(user);
                    NotificationHelper::SendToast(user);
                }
            }
            catch (hresult_error const&)
            {
            }
        }
        deferral.Complete();
    }
} // namespace winrt::TsinghuaNetHelper::implementation

File: TsinghuaNetUWP.Background/LoginTask.vb
Imports Berrysoft.Tsinghua.Net
Imports TsinghuaNetUWP.Helper
Imports Windows.ApplicationModel.Background

Public NotInheritable Class LoginTask
    Implements IBackgroundTask

    Private settings As New SettingsHelper

    Public Async Sub Run(taskInstance As IBackgroundTaskInstance) Implements IBackgroundTask.Run
        Dim deferral = taskInstance.GetDeferral()
        Try
            Dim tuple = SettingsHelper.GetInternetStatus()
            Dim state = settings.SuggestNetState(tuple.Status, tuple.Ssid)
            Dim un = settings.StoredUsername
            Dim pw = CredentialHelper.GetCredential(un)
            Dim helper = ConnectHelper.GetHelper(state, un, pw)
            If helper IsNot Nothing Then
                Using helper
                    Await helper.LoginAsync()
                    Dim user As FluxUser = Await helper.GetFluxAsync()
                    NotificationHelper.UpdateTile(user)
                    NotificationHelper.SendToast(user)
                End Using
            End If
        Finally
            deferral.Complete()
        End Try
    End Sub
End Class
